runtime/jailer.go

```diff
@@ -0,0 +1,478 @@
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, int(rdev)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	if info, err := os.Stat(dst); err == nil {
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	} else if os.IsNotExist(err) {
+		// this occurs when a full path is provided including the file name
+	} else if err != nil {
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+func chownr(path string, uid, gid int) error {
+	return filepath.Walk(path, func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		return os.Chown(name, uid, gid)
+	})
+}
+
+func (j runcJailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.Command(j.Config.RuncBinaryPath, "run", containerName)
+	cmd.Dir = j.OCIBundlePath()
+	return cmd
+}
+
+// overwriteConfig will set the proper default values if a field had not been set.
+func (j runcJailer) overwriteConfig(cfg *Config, socketPath, configPath string) error {
+	spec := specs.Spec{}
+	configBytes, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(configBytes, &spec); err != nil {
+		return err
+	}
+
+	spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+
+	spec.Root.Path = rootfsFolder
+	spec.Root.Readonly = false
+	spec.Process.User.UID = uint32(j.Config.UID)
+	spec.Process.User.GID = uint32(j.Config.GID)
+	configBytes, err = json.Marshal(&spec)
+	if err := ioutil.WriteFile(configPath, configBytes, 0666); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setDefaultConfigValues will process the spec file provided and allow any
+// empty/zero values to be replaced with default values.
+func (j runcJailer) setDefaultConfigValues(cfg *Config, socketPath string, spec specs.Spec) specs.Spec {
+	if spec.Process != nil {
+		if spec.Process.Args == nil {
+			cmd := firecracker.VMCommandBuilder{}.
+				WithBin(cfg.FirecrackerBinaryPath).
+				WithSocketPath(socketPath).
+				Build(context.Background()) // TODO: use shim context here
+
+			spec.Process.Args = cmd.Args
+		}
+	}
+	return spec
+}
+
+// getUIDAndGID will the next free uid and gid that isn't being used by the
+// system.
+//
+// TODO: This can be removed once we jail with a user namespace
+func getUIDAndGID(parentCtx context.Context, config *Config) (uint32, uint32, error) {
```

Comment:
I get that Firecracker recommends you run each VM with a different uid/gid, which I agree is a good practice, but does their jailer actually choose a uid/gid for you? Or does their jailer just ask for a uid/gid?

My concern is that because we can't actually coordinate between shim processes, the end result of the code here is actually going to be that all firecracker VMs are running with the same uid/gid (the first unused one found) unless they always override the default uid/gid ranges to be non-overlapping. So even though we are attempting to help users run with a unique uid/gid, the end result is actually the opposite in the default case.

That's on top of the fact that checking whether `/etc/passwd` and `/etc/groups` contains the uid/gid is more of a best-effort check if the uid/gid is already in use on the system, not a perfectly robust one. That's really just a limitation of the underlying OS, not something we can do anything about AFAIK.

What would be wrong with just asking for a uid/gid pair from the user and relying on them validate it's unique to the VM if they want maximum security? Obviously it's nice to take care of details for users when possible, but I'm not sure it actually is possible in this case.

Reply:
I am not opposed to it, but changing it where you require `JailerConfig` to be present means that either `JailerConfig` is a required field or jailing is disabled by default. I think I am going to make jailing disabled by default, as it provides less magic and it simplifies a lot of the code.